### ISSUE TITLE
discussion: should we keep the geo_type in the epi_df metadata or bring it over to external functions?

### ISSUE THREAD

**Issue by dshemetov** [cmu-delphi/epiprocess#523]
This is a discussion issue for determining if we want to maintain the geo_type metadata inside an epi_df. 

The alternative is extracting the code that recognizes geo_values and assigns geo_type to them and using it in a separate set of helper functions for geo-aggregation (there's even the question of whether _this_ is the right package for geo_aggregation utilities rather than something dedicated). 

Some cons of the current approach is:
- no code decisions are currently made based on geo_type, it's purely informational
- it adds maintenance work in various aggregation operations or epi_df operations (e.g. merge) where you have to figure out how to decide whether the geo_type can be maintained or it needs to be modified

Some pros of the alternative are removing the difficulties above. 

Note that removing the geo_type metadata from epi_df doesn't mean not supporting geo aggregation operations - we can either make the geo_type guess determination upon geo aggregation or have the user specify it (like in epidatpy), rather than requiring this guess to happen at every epi_df construction time.

**Comment by brookslogan** [comment]
Our validation is also not robust, both for geo_type and time_type
```r
bind_rows(as_epi_df(tibble(geo_value = "ca", time_value = as.Date("2020-01-01") + c(0,7), value = 1)), as_epi_df(tibble(geo_value = "hhs1", time_value = as.Date("2020-01-01") + c(0,1), value = 1)))
#> An `epi_df` object, 4 x 3 with metadata:
#> * geo_type  = state
#> * time_type = week
#> * as_of     = 2024-10-04 16:57:24.10385
#> 
#> # A tibble: 4 × 3
#>   geo_value time_value value
#> * <chr>     <date>     <dbl>
#> 1 ca        2020-01-01     1
#> 2 ca        2020-01-08     1
#> 3 hhs1      2020-01-01     1
#> 4 hhs1      2020-01-02     1
```
When I last looked at this, the solution to maintaining info about the geo_type would be to do it via the class of geo_value, not via the epi_df metadata.  But I'm currently taking another look at a "proper" usage of `dplyr_extending` and maybe some easier idea will come out of that.

Development effort:
- Even imperfect maintenance of the `geo_type` requires code.
- But keeping `geo_type` logic in its current places is to help detect bugs that might also be shared by `time_type` logic so we can create something solid there, and to avoid creating two different ways of handling metadata inference/maintenance that makes interface/maintenance trickier.  Fixing `time_type` maintenance might just also provide a fix for `geo_type` maintenance.
- (.... not sure if this overall favors keeping or changing.  Perhaps keeping, since it's also effort to implement any change.)

We currently use the `geo_type` in:
- Printing.  Perhaps that's useful if someone were to read something in and not know what county FIPS codes were, or if there is ambiguity in different geocoding systems (county FIPS, HRRs, state FIPS, ........).  But it's also annoying if `geo_type` inference doesn't guess the right thing, or validation refuses to accept the right thing.
  - (A wash / slightly favors changing.)

What if we actually used the `geo_type`?
- In plotting?  If there's ambiguity / bad guesses, then user will need to provide a hint/override in an argument.  Could be annoying to have to specify it again and again for multiple plots.
  - (Slightly favors keeping it the way it is.)
- In aggregation?  I'm not sure this would ever be enough... we often need associated time-dependent population crosswalks / weighted crosswalks, and they should match what the user is expecting/using, but different users could be using different things.  I guess we could put a lot of effort into some sort of sane default for common geo_types.  But I'd assume we're going to start simple and require user to pass crosswalks, so the geo_type isn't even going to be used.
  - (We're free to do whatever.)

In summary....
- I'm uncertain until I look a bit more at `dplyr_extending`
- Currently feel wary about it unless we also change how we manage `time_type`s to be similar or be `time_value`-class-based.  (I think these two approaches might actually look somewhat similar, vs. trying to store this in metadata.)

**Comment by brookslogan** [comment]
#501 discussion supports making this change.
